lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2256,6 +2256,120 @@ ClangModuleUnit *ClangImporter::Implementation::getClangModuleForDecl(
   return getWrapperForModule(M);
 }
 
+// TODO: Consolidate with importSourceLoc below
+SourceLoc ClangImporter::Implementation::convertSourceLocation(
+    const clang::SourceLocation &loc) {
+  return getBufferImporterForDiagnostics().resolveSourceLocation(
+      getClangASTContext().getSourceManager(), loc);
+}
+
+bool ClangImporter::Implementation::setDiagnosticTargetIfNotAlreadyReported(
+    const SwiftLookupTable::SingleEntry &target) {
+  bool shouldReset = false;
+  bool alreadyReported = reportedDiagnosticTargets.count(target);
+  if (!alreadyReported && !getEagerImportActive()) {
+    shouldReset = setDiagnosticTarget(target);
+    reportedDiagnosticTargets.insert(target);
+  }
+  return !alreadyReported && !getEagerImportActive() && shouldReset;
+}
+
+bool ClangImporter::Implementation::setDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &target) {
+  assert(target);
+  // In some situations, the ClangImporter is invoked in a reentrant manner,
+  // resulting in a diagnosticTarget already being set. In such a case, we
+  // keep the outer diagnostic target.
+  if (!diagnosticTarget) {
+    diagnosticTarget = target;
+    return true;
+  }
+  return false;
+}
+
+SwiftLookupTable::SingleEntry
+ClangImporter::Implementation::getDiagnosticTarget() {
+  return diagnosticTarget;
+}
+
+void ClangImporter::Implementation::resetDiagnosticTarget(bool shouldReset) {
+  if (shouldReset) {
+    assert(diagnosticTarget);
+    diagnosticTarget = nullptr;
+  }
+}
+
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &targetCandidate) const {
+  return diagnosticTarget == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::NamedDecl *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::NamedDecl *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::ModuleMacro *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::ModuleMacro *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::MacroInfo *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::MacroInfo *>() == targetCandidate;
+}
+
+bool ClangImporter::Implementation::setEagerImportActive() {
+  bool shouldReset = !eagerImportActive;
+  eagerImportActive = true;
+  return shouldReset;
+}
+
+void ClangImporter::Implementation::resetEagerImportActive(bool shouldReset) {
+  if (shouldReset) {
+    eagerImportActive = false;
+  }
+}
+
+bool ClangImporter::Implementation::getEagerImportActive() {
+  return eagerImportActive;
+}
+
+void ClangImporter::Implementation::reportNoteConsumingDiagnosticError(
```

Comment:
Is it important to call out that this is consuming a diagnostic error? This is an x-value so that will be enforced by the compiler anyway. 

The more interesting part of this function is that it only reports a note if the targets match. So, maybe a better name would be `reportNoteIfTarget` or `reportNoteForTarget` or `reportNoteIfTargetsMatch` or a better name that you come up with :)

Reply:
👍